oak_runtime/introspection_browser_client/index.tsx

```diff
@@ -45,6 +45,104 @@ function loadSerializedEvents(): Promise<Uint8Array> {
   });
 }
 
+interface OakApplicationState {
+  nodeInfos: NodeInfos;
+  channels: Channels;
+}
+
+type NodeId = number;
+type AbiHandle = number;
+type NodeInfos = Map<NodeId, NodeInfo>;
+interface NodeInfo {
+  abiHandles: Map<AbiHandle, ChannelHalf>;
+}
+
+type ChannelID = number;
+enum ChannelHalfDirection {
+  Read,
+  Write,
+}
+interface ChannelHalf {
+  channelId: ChannelID;
+  direction: ChannelHalfDirection;
+}
+interface Message {
+  data: Uint8Array;
+  channels: ChannelHalf[];
+}
+interface Channel {
+  id: ChannelID;
+  messages: Message[];
+}
+type Channels = Map<ChannelID, Channel>;
+
+function eventReducer(
+  applicationState: OakApplicationState,
+  event: introspectionEventsProto.Event
+) {
+  const eventType = event.getEventDetailsCase();
+  const { EventDetailsCase } = introspectionEventsProto.Event;
+
+  switch (eventType) {
+    case EventDetailsCase.NODE_CREATED:
+      applicationState.nodeInfos.set(event.getNodeCreated().getNodeId(), {
+        abiHandles: new Map(),
+      });
+
+      break;
+    case EventDetailsCase.NODE_DESTROYED:
+      applicationState.nodeInfos.delete(event.getNodeDestroyed().getNodeId());
+
+      break;
+    case EventDetailsCase.CHANNEL_CREATED:
+      {
+        const channelId = event.getChannelCreated().getChannelId();
+        applicationState.channels.set(channelId, {
+          id: channelId,
+          messages: [],
+        });
+      }
+
+      break;
+    case EventDetailsCase.CHANNEL_DESTROYED:
+      applicationState.channels.delete(
+        event.getChannelDestroyed().getChannelId()
+      );
+
+      break;
+    case EventDetailsCase.HANDLE_CREATED:
+      {
+        const details = event.getHandleCreated();
+        applicationState.nodeInfos
+          .get(details.getNodeId())
+          .abiHandles.set(details.getHandle(), {
+            channelId: details.getChannelId(),
+            direction: 0,
+          });
+      }
+      break;
+    case EventDetailsCase.HANDLE_DESTROYED:
+      {
+        const details = event.getHandleDestroyed();
+        applicationState.nodeInfos
+          .get(details.getNodeId())
+          .abiHandles.delete(details.getHandle());
+      }
+      break;
+    case EventDetailsCase.MESSAGE_ENQUEUED:
+      // TODO(#913): Add support for displaying messages
+      break;
+    case EventDetailsCase.MESSAGE_DEQUEUED:
+      // TODO(#913): Add support for displaying messages
```

Comment:
At the moment it is not clear to me how the state, including where channel-falves are stored will be fully reconconstructed. Will this event include a list of channel halves that was passed in the message?

Reply:
Yup! See the associated proto of this event: 

https://github.com/project-oak/oak/blob/c3d9e43aeab28095643788d1e2e91f984d31ddfe/proto/introspection_events.proto#L82